### ISSUE TITLE
PM2 is stuck in restart loop when watching a folder

### ISSUE THREAD

**Issue by rbhalla** [Unitech/PM2#1243]
I am using a mounted nfs folder to share my app directory from my local machine to a vagrant machine. For this reason I enable `usePolling` with chokidar when enabling watch.

For some reason when I do this pm2 gets stuck in a loop restarting itself because of an apparent change, like so:

```
PM2: 2015-05-07 11:12:07: Starting execution sequence in -cluster mode- for app name:app id:0
PM2: 2015-05-07 11:12:07: App name:app id:0 online
PM2: 2015-05-07 11:12:09: Change detected for app name: app - restarting
PM2: 2015-05-07 11:12:09: Stopping app:app id:0
PM2: 2015-05-07 11:12:09: App name:app id:0 disconnected
PM2: 2015-05-07 11:12:09: App name:app id:0 exited with code 0
PM2: 2015-05-07 11:12:09: Process with pid 7896 killed
PM2: 2015-05-07 11:12:09: Starting execution sequence in -cluster mode- for app name:app id:0
PM2: 2015-05-07 11:12:09: App name:app id:0 online
PM2: 2015-05-07 11:12:10: Change detected for app name: app - restarting
PM2: 2015-05-07 11:12:10: Stopping app:app id:0
PM2: 2015-05-07 11:12:10: App name:app id:0 disconnected
PM2: 2015-05-07 11:12:10: App name:app id:0 exited with code 0
PM2: 2015-05-07 11:12:10: Process with pid 7910 killed
```

I created a gulp file and using chokidar watched for changes (using the same watch options) and this didn't occur. Nothing in `Watcher.js` really strikes me as immediately wrong so I am unable to determine the cause of this.

I was wondering if anyone else might be able to provide some insight?

**Comment by soyuka** [comment]
It seems that when launching, your app is writing some file. Is it logging or anything like that on startup?

**Comment by rbhalla** [comment]
You're right, I didn't do enough digging before posting this, apologies.

My issue was directly related with how sails works so apologies for polluting the issues list here. I'll post how I fixed it for anyone wandering across this in the future.

In this scenario I was using sails js. When a server is started, sails automatically compiles everything. You need to therefore add `.tmp` to your ignore list. If for some reason you're doing this with a production environment, you should also add the `www` folder to your ignore list.

Sails also (by default) injects js/css script locations directly into your layout files. This will always cause an update event regardless of whether the file was changed. It's up to you whether you keep it, but if you do you need to ignore your `layouts` folder. Or better yet disable this completely.

**Comment by rbhalla** [comment]
Ok, this issue is still actually occurring for me but not as persistently as before.

It's actually quite easy to produce. Use the express generator to create a bare bones express app, then use the following json file to start it:

```
{
    "apps": [{
        "name": "api",
        "script": "bin/www",
        "cwd": "/var/www/html",

        "instances": 1,
        "autorestart": true,
        "watch": "../",
        "ignore_watch" : ["node_modules", "bower_components", ".git"],
        "watch_options": {
            "usePolling": true
        }
    }]
}
```

**Comment by rbhalla** [comment]
Ok some more digging and I've found the problem and the solution. Although I'm not quite sure what caused this problem.

It seems like the file being changed is `/home/deploy/.pm2/pm2.log`. I am not quite sure why `../` would cause that directory to be watched, but the fix was to add the full path to the watch string. In the above case that would be `/var/www/html`.

If anyone could tell me why `../` would make that directory be watched it'd be helpful to me understanding it all.

**Comment by gihrig** [comment]
Similar issue with pm2 reporting an app is "stopping" but it never stops. 
Everything worked as expected the first day after install and development of script files with many start, stop, ls, and few del commands.

I tried deleting all files from `~/.pm2/pids`, `pm2 -f del {$app}, pm2 kill`.
There is no `pm2` or `node` process visible in Activity Manager.

None of the above cleared the app with the stuck `stopping` status.

I finally cleared the problem by removing and reinstalling pm2.

In my case the problem app is a Sapper app started from the cli via bash script.

````
    pm2 start ~/$app/package/target/release/$app --watch --name $app
    pm2 start -f ~/$app/package/__sapper__/build --watch --name $app-ui
````

(The app on the fist line is an OS executable, compiled rust app)

````
macOS 10.14.6
node: 12.18.3
pm2: 4.5.0
````

pm2 was installed with `sudo npm i -g pm2`
